Kokkos backend with minimal dependencies on vendor APIs

cwsmith opened the pull request:
This PR adds support for a Kokkos backend with minimal dependencies on vendor APIs (i.e., AMD HIP, NVIDIA CUDA, Intel SYCL/DPL).  All tests are passing with HIP 5.5.1 on Frontier (see https://github.com/SCOREC/omega_h/wiki/Build-and-Run-on-OLCF-Frontier for details) and on NVIDIA GPU systems with various versions of CUDA, up to and including CUDA 12.1 on Perlmutter.

The only use of vendor libraries is for stable sorting on GPUs.

An implementation of a Kokkos-based memory pool is also included.  Details can be found [here](https://github.com/matthew-mccall/omega_h-kokkos-mempool-paper).

`master` will be tagged `scorec-v10.7.0` once this is merged.

joshia5 commented on src/Omega_h_array.hpp:
```diff
@@ -47,61 +65,20 @@ class Write {
   std::string name() const;
 #else
-  std::string const& name() const;
+      std::string const& name() const;
```
whitespace

cwsmith replied:
Thank you. Fixed with [8f350da](https://github.com/SCOREC/omega_h/pull/60/commits/8f350da36549a245e0b3c26c699c7db2eaf1ff6d).

joshia5 commented on src/Omega_h_array_kokkos.hpp:
- [ ] fix indentation
@cwsmith a general question, will it be a short or long term thing to move all function definitions that have vendor/kokkos specific calls into .cpp files from .hpp?

cwsmith replied:
I don't think there was too much code reorganization outside of the mempool (more on that below).  Is there a specific file/change that caught your attention?

In the long term the plan is to remove the CUDA and OpenMP backends.  After that, the only location we will have vendor API calls will be in the sorting implementation (`src/Omega_h_sort.cpp`) (assuming Kokkos doesn't provide stable sort at that time).

Note, adding the kokkos mempool support had the most significant reorganization change as we wanted to preserve the CUDA based pool and avoid lots of preprocessor conditionals (`#ifdef ...`).  Hence, we now have `Omega_h_array_default.hpp` and `Omega_h_array_kokkos.hpp` which are switched [here](https://github.com/SCOREC/omega_h/blob/8f350da36549a245e0b3c26c699c7db2eaf1ff6d/src/Omega_h_array.hpp#L208-L212).

cwsmith replied:
indentation fixed with e85a219

joshia5 commented on src/Omega_h_fail.hpp:
- [ ] fix whitespaces

cwsmith replied:
@joshia5 The whitespace in all the pre processor macros is relatively consistent where changes were made.  Is there something specific that was concerning?

joshia5 commented on src/Omega_h_map.cpp:
```diff
@@ -172,6 +174,7 @@ Read<I8> invert_marks(Read<I8> marks) {
 }
 
 LOs collect_marked(Read<I8> marks) {
+  static int callNum=0;
```
was this added for debugging?

cwsmith replied:
Good catch.  Fixed with https://github.com/SCOREC/omega_h/pull/60/commits/7e19b3c62fd6d36ce703ed599964306f0ad30fd5.

joshia5 commented on src/Omega_h_memory.hpp:
- [ ] fix indentation (mixture of 4 space and 2 space indentation scheme in multiple places)

cwsmith replied:
I don't see an obvious problem with indentation.  This may have be an artifact of github rendering if that is where you were looking at the changes.

joshia5 commented on src/Omega_h_pool_kokkos.hpp:
```diff
@@ -0,0 +1,122 @@
+constexpr size_t DEFAULT_BYTES_PER_CHUNK = 1000;
+
+// Total initial size in bytes will subsequently be DEFAULT_NUMBER_OF_CHUNKS * DEFAULT_BYTES_PER_CHUNK
+constexpr size_t DEFAULT_NUMBER_OF_CHUNKS = 700'000;
+
+// If the largest StaticKokkosPool is 2Kb, then the next StaticKokkosPool will be at least 4kB resulting in a total pool size of 2Kb + 4Kb = 6Kb
+constexpr size_t DEFAULT_GROWTH_CONSTANT = 2;
```
@cwsmith if the initial pool size has enough of an effect on compile and run time, something to think about is try and relate it to initial mesh size.
Another thing: to have memory in multiples of 2^X,might be useful

cwsmith replied:
runtime vs initial pool size: Matthew's testing on the DeltaWing cases (see https://github.com/matthew-mccall/omega_h-kokkos-mempool-paper/) indicates that the cost to resize when the initial pool size is 'small' is minimal (less than 2-3% it seems) relative to a 'large' (using a-pirori knowledge of memory usage) initial pool.

2^x sizing: Thank you.  That may help.  We are looking at some other approaches and trying to see if there are other opportunities like this.

joshia5 commented on src/Omega_h_array.hpp:
```diff
@@ -47,57 +65,16 @@ class Write {
   Write(LO size_in, T offset, T stride, std::string const& name = "");
   Write(std::initializer_list<T> l, std::string const& name = "");
   Write(HostWrite<T> host_write);
-  OMEGA_H_INLINE LO size() const OMEGA_H_NOEXCEPT {
```
dont the array default and arraykokkos headers need to include array.hpp

matthew-mccall replied:
No, it's actually the other way around, array.hpp includes arraydefault or arraykokkos. arraydefault and arraykokkos provide the out-of-line definitions for the templated Write class. Since the definitions differ between the existing backend and the Kokkos backend, I saw it fit to separate these out-of-line definitions into two separate header files to reduce on the preprocessor directives in an effort to make the code more readable. These definitions must be in a header file as Write is templated, and after declaration of the Write class. As such they are included by array.hpp at the end of file since the #include preprocessor directive in effect "copies and pastes" the contents of the file into the spot where the include is, thus satisfying the requirements of out-of-line definitions of a templated class.

joshia5 commented on src/initKokkosAndLib.cpp:
this test should be sufficient by itself or do we think both this test and the initKokkos.cpp are helpful?

cwsmith replied:
Good catch. The `initKokkos` test was removed with d35a284.

joshia5 commented on src/unit_array_algs.cpp:
```diff
@@ -77,9 +103,21 @@ static void test_scan() {
+    fill<I8>(marks_w,1);
+    auto offsets = offset_scan(Read<I8>(marks_w));
+    parallel_for(offsets.size(), OMEGA_H_LAMBDA(LO i) {
+        assert(offsets[i] == i);
```
indentation

cwsmith replied:
fixed in 1eb963e

joshia5 commented:
Some additional comments which can be helpful to have :
- [ ] a list of any changes in either behavior or arguments/calls that are expected from user side to keep existing things backwards compatible. Though i think the regression testing will have checked for most of this
- [ ] a ctest/util containing simple few line of comments on how to run a case with option of pool on/off

cwsmith commented:
> Some additional comments which can be helpful to have :
> 
>     * [ ]  a list of any changes in either behavior or arguments/calls that are expected from user side to keep existing things backwards compatible. Though i think the regression testing will have checked for most of this
> 
>     * [ ]  a ctest/util containing simple few line of comments on how to run a case with option of pool on/off

- Since it was just a change in how the kokkos backend works, the user experience is identical.
- The cmake option `-DENABLE_CTEST_MEMPOOL` will enable testing with the mem pool enabled.  Passing `--osh-pool` to a binary linked against omegah will enable the pool.

cwsmith commented:
@joshia5 Thank you for the review.  I think we addressed all the comments except maybe a few on whitespace/indentation.  Would you please take another look to confirm?

joshia5 reviewed (approved):
seems good to go
